### DISCUSSION TITLE
Custom middleware, adding JSON parser

### DISCUSSION THREAD

**Discussion by AgrYpn1a** [scottoffen/grapevine#117] [Q&A] [answered]
Hey, is it possible to add custom middleware? In my example, I need to parse the request body as JSON, so I don't wanna do it every time in each route, would be great if it can be included as a middleware and an extra string field where it contains a JSON that can be parsed from there, or even a generic parameter specifier that can be passed via attributes or something, so that the body is read as the specified type directly.

Thanks!

**Comment by scottoffen** [comment, marked as answer]
Yes.

You can parse the JSON in a single place for all your route using `Router.BeforeRoutingAsync`. These handlers are delegates of type `RoutingAsyncEventHandler`, and run just before any routes are executed (and there is a corresponding `router.AfterRoutingAsync`). This is where you want add your middleware.

In your middleware, after parsing the body, you can add the json to the `IHttpRequest.Locals` property, a key-value store that is used to pass state between all routes for a single request.

Here is one sample of how to do this:

```csharp
public static class IRestServerExtensions
{
    public static void AutoParseJson(this IRestServer server)
    {
        server.Router.BeforeRoutingAsync += ParseJsonAsync;
    }

    public static async Task ParseJsonAsync(IHttpContext context)
    {
        // Only attempt to parse if the Content-Type header is application/json
        // and there is content in the request body
        if (context.Request.ContentType != ContentType.Json || !context.Request.HasEntityBody) return;

        // Parse your json body using whatever library you choose.
        var json = await JsonSerializer.DeserializeAsync<object>(context.Request.InputStream);

        // Attached the parsed data to the request locals
        context.Locals.TryAdd("your-json-key", json);
    }
}
```

You can add this middleware via `server.AutoParseJson()`, and then your parsed JSON will now be available in all of your routes by referencing `context.Locals.Get("your-json-key");`.

The places you can attach middleware are:
- `IRestServer.OnBeforeStarting`
- `IRestServer.OnAfterStarting`
- `IRestServer.OnBeforeStopping`
- `IRestServer.OnAfterStopping`
- `IRestServer. OnRequestAsync`
- `IRouter.BeforeRoutingAsync`
- `IRouter.AfterRoutingAsync`

**Reply by scottoffen** [reply]
Middleware for JSON serialization isn't included by default because I've worked on so many projects over the years that were bringing in conflicting dependencies on JSON serializers, and saw the hoops they had to jump through to get things to work, which also made debugging a pain.

This approach allows you to serialize and deserialize using your method/library of choice.

**Reply by AgrYpn1a** [reply]
Thanks a lot :) While this works, I'd have to deserialize again in each specific route, to get the type I need, is there a way to make this more generic? (Type casting doesn't work btw)

**Reply by scottoffen** [reply]
Not deserialize, but cast or map from the JSON object to your desired object. I've been thinking about this quite a bit recently, and I don't have any good ideas that would make it simpler yet. I'm open to suggestions.

I realize this is built-in to ASP.NET, but that is largely due to the approach they take with how they route requests - there is only ever one route (controller) that you will hit and you bind a model to it - which is very different from how Grapevine works.

**Reply by AgrYpn1a** [reply]
Okay, thanks man! :) I'll let you know if I have any suggestions, but for now the solution I have is good enough,
